### PR TITLE
feat(offchain): add transfer from ckb

### PR DESCRIPTION
**Pull request by fpChan**
- [x]  add transfer from ckb

- [x]  test unlock erc20 token

### CODE REVIEW

File: eth-contracts/address.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (huwenchao) [inline eth-contracts/address.json:1]: should remove this file and add it into .gitignore

File: eth-contracts/scripts/geth/deployAll.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+  //ERC20
+  const ERC20 = await ethers.getContractFactory(
+    "contracts/test/ERC20.sol:ERC20"
+  );
+  const ERC20Deploy = await ERC20.deploy();
+  await ERC20Deploy.deployed();
+  const ERC20DeployAddr = ERC20Deploy.address;
+  console.log("ERC20 deployed to:", ERC20DeployAddr);
+  console.log("waiting for block confirmations, about 1 minute");
+  await sleep(60 * 1000);
```
Comment (huwenchao) [inline eth-contracts/scripts/geth/deployAll.js:25]: do not need to wait

File: eth-contracts/scripts/geth/deployAll.js [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+  const contractBalance = await ERC20Deploy.callStatic.balanceOf(
+    "0x17c4b5CE0605F63732bfd175feCe7aC6b4620FD2"
```
Comment (huwenchao) [inline eth-contracts/scripts/geth/deployAll.js:28]: what is this address stands for?

File: eth-contracts/scripts/geth/deployAll.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    "0x94e9522537032ef62fd21043ae1b1247738ffe264b16392ea5e2fe752376b5e0";
+  let res = await CKBChinDeploy.initWithHeader(
+    initHeaderData,
+    initBlockHash,
+    finalizedGcThreshold,
+    canonicalGcThreshold
+  );
+  console.log("initHeader :", res.hash);
+  console.log("waiting for block confirmations, about 1 minute");
+  await sleep(60 * 1000);
```
Comment (mkxbl) [inline eth-contracts/scripts/geth/deployAll.js:61]: delete this or if you want to wait for confirmations, use `tx.wait(confirmations)`

File: eth-contracts/scripts/geth/deployAll.js [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+  // deploy Migrations
+  const Migrations = await ethers.getContractFactory(
+    "contracts/Migrations.sol:Migrations"
+  );
+  const MigrationsDeploy = await Migrations.deploy();
+  await MigrationsDeploy.deployed();
+  const MigrationsDeployAddr = MigrationsDeploy.address;
+  console.log("Migrations deployed to:", MigrationsDeployAddr);
+  console.log("waiting for block confirmations, about 1 minute");
+  await sleep(60 * 1000);
```
Comment (mkxbl) [inline eth-contracts/scripts/geth/deployAll.js:83]: no need to deploy migrations

File: eth-contracts/scripts/geth/deployAll.js [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+  const finalizedGcThreshold = 500;
+  const canonicalGcThreshold = 40000;
+
+  const initHeaderData =
+    "0x0000000000000820a9daa6d47501000036100000000000000200004605080700b3f696ffbda4feb72731b797a135c5cc779200acea4aa45cac7df07e856403f350f98d988b232747c7f757a4b48b733754d063101934fe69a4128b5ad25855a300000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000de4a62557a97a32ef49cf85a678723000df76dee6b82000000e9fbaead700a0741571b602446aa2a2135177404285c15";
+  const initBlockHash =
+    "0x94e9522537032ef62fd21043ae1b1247738ffe264b16392ea5e2fe752376b5e0";
```
Comment (mkxbl) [inline eth-contracts/scripts/geth/deployAll.js:52]: add a command to init ckb light client

File: offchain-modules/lib/src/transfer/to_eth.rs [hunk 7]
```diff
@@ -60,6 +60,45 @@ pub fn burn(
...
+) -> Result<()> {
+    let mut ckb_client = HttpRpcClient::new(ckb_url);
+    let hash = covert_to_h256(&tx_hash)?;
+    let block_hash = ckb_client
+        .get_transaction(hash)
+        .map_err(|err| anyhow!(err))?
+        .ok_or_else(|| anyhow!("tx is none"))?
+        .tx_status
+        .block_hash
+        .ok_or_else(|| anyhow!("block hash is none"))?;
```
Comment (mkxbl) [inline offchain-modules/lib/src/transfer/to_eth.rs:78]: should sleep and retry if get block_hash error

File: offchain-modules/quick_start.md [hunk 8]
```diff
@@ -0,0 +1,305 @@
+## 环境准备
```
Comment (huwenchao) [inline offchain-modules/quick_start.md:1]: rename file to `quick_start_CN.md`

File: eth-contracts/scripts/geth/deployAll.js (hunk 3 above)
Comment (fpChan) [reply eth-contracts/scripts/geth/deployAll.js]: > what is this address stands for?

the erc20 contract deployer, just for test.  It been removed

File: docker/README.md [hunk 9]
```diff
@@ -24,6 +24,8 @@ docker run --rm -it -p 8116:8116 ckb-indexer
 
 # Geth private chain
 
+
+
```
Comment (mkxbl) [inline docker/README.md:28]:
```suggestion
```

File: eth-contracts/scripts/geth/deployAll.js [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+  const data = JSON.stringify(address);
+
+  fs.writeFile("./scripts/geth/address.json", data, (err) => {
+    if (err) {
+      throw err;
+    }
+    console.log("JSON data is saved.");
+  });
+  await sleep(60 * 1000);
```
Comment (mkxbl) [inline eth-contracts/scripts/geth/deployAll.js:75]: why

File: offchain-modules/cli/src/commands/mod.rs [hunk 11]
```diff
@@ -23,6 +24,7 @@ use web3::types::{H256, U256};
 
 pub async fn handler(opt: Opts) -> Result<()> {
     match opt.subcmd {
+        SubCommand::InitLightContract(args) => init_light_contract_handler(args).await,
```
Comment (mkxbl) [inline offchain-modules/cli/src/commands/mod.rs:27]:
```suggestion
        SubCommand::InitCkbLightContract(args) => init_ckb_light_contract_handler(args).await,
```

File: offchain-modules/lib/src/transfer/to_eth.rs [hunk 12]
```diff
@@ -60,6 +97,59 @@ pub fn burn(
...
+            .ok_or_else(|| anyhow!("tx is none"))?
+            .tx_status
+            .block_hash;
+        match block_hash_opt {
+            Some(hash) => {
+                block_hash = hash;
+                break;
+            }
+            None => {
+                info!("the transaction is not in block yet");
```
Comment (mkxbl) [inline offchain-modules/lib/src/transfer/to_eth.rs:124]:
```suggestion
                info!("the transaction is not committed  yet");
```

File: offchain-modules/quick_start_CN.md [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+  ```
+
+  ```
+  {
+    "ERC20Deploy": "0xE9837e999934E2f61D4148CeEEb1Be3cc00Cb37c",
+    "CKBChainDeploy": "0x3C049b35392D20e5C1E7bB6B32E26f07093A6923",
+  // "EaglesongDeploy": "0x4414f183026502850203cE9df08f2648E1C39b32",
+  // "MigrationsDeploy": "0xBE2AB31F46e740f15d93331962f930d39d79920c",
+    "TokenLockerDepoly": "0x30F58CEEEC7Ace3fC2f938F13dD2C0c69B4A19b0"
+  }
```
Comment (mkxbl) [inline offchain-modules/quick_start_CN.md:40]:
```suggestion
  {
    "ERC20Deploy": "0xE9837e999934E2f61D4148CeEEb1Be3cc00Cb37c",
    "CKBChainDeploy": "0x3C049b35392D20e5C1E7bB6B32E26f07093A6923",
    "EaglesongDeploy": "0x4414f183026502850203cE9df08f2648E1C39b32",
    "TokenLockerDepoly": "0x30F58CEEEC7Ace3fC2f938F13dD2C0c69B4A19b0"
  }
```

File: eth-contracts/scripts/geth/deployAll.js (hunk 10 above)
Comment (fpChan) [reply eth-contracts/scripts/geth/deployAll.js:75]: > why sleep

the func is async type.

File: docker/README.md [hunk 14]
```diff
@@ -36,9 +38,19 @@ The first time you run container, it will take 5~10 minutes for geth to be ready
...
 
+The privkey of accounts in geth-genesis.json for tests:
+
+| pubkey | privkey |
+| :----: | :-----: |
+| 0x17c4b5CE0605F63732bfd175feCe7aC6b4620FD2 | 0xc4ad657963930fbff2e9de3404b30a4e21432c89952ed430b56bf802945ed37a |
+| 0x8951a3DdEf2bB36fF3846C3B6968812C269f4561 | 0x719e94ec5d2ecef67b5878503ffd6e1e0e2fe7a52ddd55c436878cb4d52d376d |
+| 0x42e8763917A72e07369AD54B158b0FA839f060bc | 0x627ed509aa9ef55858d01453c62f44287f639a4fa5a444af150f333b6010a3b6 |
+| 0xE61438B717b6937388bf66D256395A15B3D169aE | 0x49e7074797d83cbb93b23877f99a8cecd6f79181f1236f095671017b2edc64c2 |
+
```
Comment (mkxbl) [inline docker/README.md:51]:
```suggestion
The pre-mined accounts in geth-genesis.json for tests, please replace them to other accounts you have privkeys.

| pre-mined accounts |
| :----: | 
| 0x17c4b5CE0605F63732bfd175feCe7aC6b4620FD2 | 
| 0x8951a3DdEf2bB36fF3846C3B6968812C269f4561 | 
| 0x42e8763917A72e07369AD54B158b0FA839f060bc | 
| 0xE61438B717b6937388bf66D256395A15B3D169aE | 

```

File: offchain-modules/quick_start_CN.md [hunk 15]
```diff
@@ -0,0 +1,300 @@
+## 环境准备
+
+### 运行 ETH 和 CKB节点
+
+[docker 运行](https://github.com/nervosnetwork/force-bridge-eth/tree/main/docker) 即可 ，会运行 `ckb`,   `ckb-indexer`,  `eth`  三个服务，可修改节点相应配置
+
+
+### 账户说明
+
+| 私钥                                                         | 地址                                             | 用途说明                               | 修改方式                                                        |
```
Comment (mkxbl) [inline offchain-modules/quick_start_CN.md:10]: remove privkeys and add cli/privkeys to .gitignore

File: offchain-modules/quick_start_CN.md [hunk 16]
```diff
@@ -0,0 +1,299 @@
...
+    eth_contract_address: "0x30F58CEEEC7Ace3fC2f938F13dD2C0c69B4A19b0",
+    eth_token_address: "0xE9837e999934E2f61D4148CeEEb1Be3cc00Cb37c",
+}
+```
+
+
+
+3、 初始化 ckb_light_client 合约
+```shell script
+RUST_LOG=info ./target/debug/force-eth-cli init-light-contract -i 4150 --private-key-path cli/privkeys/ckb_key --to $CKBChainDeployAddr  -f 500 -c 40000
```
Comment (mkxbl) [inline offchain-modules/quick_start_CN.md:81]: describe the meaning of  -i param
```suggestion
RUST_LOG=info ./target/debug/force-eth-cli init-ckb-light-contract -i 4150 --private-key-path cli/privkeys/eth_key --to $CKBChainDeployAddr  -f 500 -c 40000
```

**Pull request closed without merging**